[release/8.0] [Blazor] Improvements to SSR component activation

MackinnonBuck opened the pull request:
# [release/8.0] [Blazor] Improvements to SSR component activation

Improves how components rendered from a Blazor endpoint get activated for interactivity.

## Description

This PR makes the following improvements:
1. An update from a Blazor endpoint cannot supply new parameters to an existing component, unless the component specifies a formattable, deterministic `@key` that uniquely identifies it.
2. The number of interactive Server root components originating from a Blazor endpoint cannot exceed the configured `CircuitRootComponentOptions.MaxJSRootComponents` option at any point in time.

Fixes #50849

## Customer Impact

This change helps to limit the quantity of server resources used by Blazor Server interactivity and ensure correctness when dynamically supplying component parameter updates.

## Regression?

- [ ] Yes
- [X] No

## Risk

- [ ] High
- [X] Medium
- [ ] Low

The change that limits the total number of interactive server root components is fairly low-risk, and the default limit of 100 interactive server root components can be increased for apps that need it.

However, the change to how new parameters are supplied is going to modify the behavior of Blazor Web apps built on previous .NET 8 preview releases. Furthermore, the code changes touch some of the core functionality of Blazor's new .NET 8 features. However, apps built prior to .NET 8 should continue to work the same way without any changes in behavior.

## Verification

- [X] Manual (required)
- [X] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [X] N/A

MackinnonBuck committed:
Dispose+re-init SSR'd components if params change

MackinnonBuck committed:
Dispose+re-init SSR'd components if params change

MackinnonBuck commented on src/Components/Components/src/ParameterView.cs:
```diff
@@ -145,10 +145,13 @@ internal bool HasDirectParameter(string parameterName)
-    // ParameterView instances, and even if there was, it's unlikely it should
-    // use these equality rules which are designed for their effect on rendering.
-    internal bool DefinitelyEquals(ParameterView oldParameters)
+    /// <summary>
+    /// Returns whether the parameters in this <see cref="ParameterView"/> are definitely equal to <paramref name="oldParameters"/>.
+    /// This is primarily used to detect whether a component's parameters have changed in a way that requires it to be re-rendered.
+    /// </summary>
+    /// <param name="oldParameters">The parameters to compare against.</param>
+    /// <returns><c>true</c> if the parameters are definitely equal, otherwise false.</returns>
+    public bool DefinitelyEquals(ParameterView oldParameters)
```
We might be able to avoid adding a public API if we added a shared implementation of this method that relied on the `ParameterView` enumerator. However, we would also have to split up part of the implementation of the internal `ChangeDetection` class so that _some_ of it is shared (it references `IEventCallback`, which is internal). Since event callbacks can't be supplied to root components, it's fine if we skip that check for root components anyway. But it's a question of whether we want to jump through all those hoops to share as much of the implementation as possible while avoiding new public API, or if it's better to just make this API public.

javiercn replied:
Don't make this public, make it shared code at this point.

MackinnonBuck commented on src/Components/Server/src/Circuits/RemoteRenderer.RootComponentOperations.cs:
```diff
@@ -0,0 +1,46 @@
+// Contains logic for root components that get dynamically updated in Blazor Web scenarios.
+internal partial class RemoteRenderer
+{
+    private Dictionary<int, ParameterView>? _latestRootComponentDirectParameters;
```
There's also `_rootComponentsLatestParameters` in `Renderer.cs`, but that's only used/updated for hot reload scenarios. I want to limit the tracking of root component parameters to precisely the scenarios that need it.

MackinnonBuck commented:
Some thoughts:

1. The gap between disposing and re-initializing a component can cause the component to flicker (particularly on Blazor Server). I see a few options for addressing this:
    1. Update the core rendering logic to allow swapping which root component instance is associated with which `ComponentState` (so the component ID, etc. gets reused), and let the diffing algorithm handle updates without the client ever knowing that it's working with a new component instance. I haven't investigated thoroughly the feasibility or drawbacks of this.
    2. Wait to clear the DOM associated with the old component until the new component takes its place. The user could still interact with the old DOM until that transition happens. I haven't yet investigated whether drawbacks come with that approach. Technically, there already is a gap between component disposal and the DOM updating accordingly, so this might just be barely extending that gap, which I don't think would be problematic.
    3. Don't do anything and improve this further in .NET 9.
2. We might be able to securely allow components to accept new component parameters:
    * We already allow specifying a `@key` to give the client a hint about how to match up components.
    * We could add the key to the data-protected payload and check on the server if the `@key` matches that of the existing component. If it does (and if the key is specified at all), we update the parameters of the component. Otherwise, we dispose and re-initialize the component if its parameters haven't changed.
    * We would have to think carefully about whether something like this could be done after .NET 8 without it being considered breaking.

What do you think? @SteveSandersonMS @javiercn

javiercn commented:
>Giving developers a mechanism for really updating parameters in place if that's important to them
Your point that we could use @key is excellent, and the more I think of it, the more important this seems. There really are valid cases for updating an interactive root's parameters in place, for example switching between pages on a grid, where each page link updates a querystring parameter.

> If the default behavior is "parameter changes mean creating a new interactive root" but people can add @key to opt into retention, that seems like it solves multiple problems at once. Once there's a @key, we interpret it as the developer promising that this root component is safe to accept any set of parameters previously rendered for a root component with the same type and key.

> Obviously some people will feel that it would be better still if they got retention without needing @key, but unless we're willing to accept the possibility that a random interactive root component could receive an unexpected set of parameters from another component instance, we wouldn't be able to do that.

> So the rule we'd have to document/explain would be:

> For normal components with parameter updates, we retain by default (but you can use @key force retention to happen or not happen)
> For components at interactive roots with parameter updates, we replace by default (but you can use @key force retention to happen or not happen)
> It seems manageable to me. But what does this do to the implementation cost?

We only need to use `@key` when there are more than one instance of a given component, isn't it? If there's only one instance, we can unambiguously pass new parameters to it. Is that not the case? 
The only potential case where this could go wrong (I think) is if the component rendered after the first instance represents a different location on the render tree.

If that is the case, then using `@key` will have the same problems, isn't it? Unless we start requiring `@key` to be unique across the document (essentially making it an id) which is a whole different game.

Will we avoid the flicker if instead we wait to remove the existing component until we have the HTML for the new component ready and we swap them in one go?

Hashing the parameters seems reasonable, but it has security implications unless we use an HMAC (we can't just SHA256 the parameters, if you want to know why ask me offline)

MackinnonBuck commented:
> After each batch of root component updates, the number of active Blazor Server components may not exceed the number of components rendered with the "auto" or "server" render modes in the last SSR update (prototyped, not yet in this PR).

I've taken a slightly different approach than what this comment says, and it's a different approach than what I prototyped as well. But I think it's equally valid and has some benefits over the other approach as well.

## How it works:
* Every component for the most recent SSR invocation has its component ID stored in a hash set
* When a component gets removed, its ID gets removed from the hash set
* When we detect that a new invocation has started, we want to guarantee that all components in the current hash set are either:
  * Updated (because they were re-rendered on the page)
  * Removed (if they were removed from the page)
* To do this, we treat the current hash set as the hash set for the previous invocation, and start a new hash set for the new invocation. By the time all update/remove operations have occurred, we expect the preivous hash set to be empty, and the new hash set to represent the new set of SSR'd interactive components.
* For each update operation, move the component from the old hash set to the new one.
* For each remove operation, remove the component from the old hash set.
* If, at this point, the old hash set still has some root components, that's a fatal error.
* For each add operation, add the new root component to the new invocation hash set.

## Benefits:
* It doesn't require round-tripping the number of expected interactive components to the DOM and back to the circuit.
* It avoids miscounting root components that get added some other way than via an SSR'd component from a Blazor endpoint (e.g., JS root components).
* It could potentially allow us to automatically dispose components that the client failed to do instead of treating as an error (if we find that to be an appealing option).

I still need to add tests to verify that all scenarios work correctly, but I've drafted some formal proofs that hopefully explain why this works (although I do admit that this area is more complicated than these proofs make it out to be).

<details>

<summary>Proof 1: The total number of root components in a circuit may not exceed the number most previously rendered via Blazor endpoint.</summary>

**Lemma 1**: In order for the number of root components to increase, an "add" operation must occur.

**Lemma 2**: If the circuit currently has 0 root components, and "add" operations cannot be duplicated, then the number of root components on the circuit cannot exceed the number of "add" operations produced via Blazor endpoint.

**Lemma 3**: If an "add" operation occurs from an unseen endpoint invocation, all existing components must first be removed, meaning the circuit must first have 0 root components.

**Lemma 4**:

Suppose the circuit started with 0 root components, and all the previous lemmas are correct.

When an add operation occurs, there are three possibilities:
1. The "add" operation comes from a previous, but not most recent endpoint invocation
2. The "add" operation comes from the most recent endpoint invocation
3. The "add" operation comes from a new endpoint invocation

In case 1, the add operation is invalid, because the circuit checks that the operation doesn't come from previously-seen invocation that isn't the most recent one. The circuit disconnects.

In case 2, as proven by lemma 2, the circuit has not exceeded the number of root components included in the most recent endpoint invocation, because:
* The circuit started at 0 root components
* The operation came from the most recent endpoint invocation
* The operation has not been duplicated

In case 3, as described by lemma 3, all existing root components must have first been removed before the new endpoint invocation occurs. If this is not the case, the circuit disconnects. If it is the case, then do not have more root components than what was rendered by the most recent Blazor endpoint.

Since all 3 possibilities result in either the circuit disconnecting or the number of circuit components being less than or equal to the number of components rendered from the most recent Blazor endpoint, then it's not possible for the number of root components in the circuit to exceed the number of root components rendered via Blazor endpoint.

QED.

</details>

<details>

<summary>Proof 2: The client never adds any root components from a new Blazor endpoint invocation before disposing all components from the previous Blazor endpoint invocation.</summary>

**Lemma 1**: New circuit root components cannot be added from a new invocation until the invocation completes.

**Lemma 2**: Remove operations happen as soon as a new invocation starts.

**Lemma 3**: Update operations may happen as soon as, or any time after, an invocation starts. Update operations are equivalent to a "remove" followed by an "add". It follows that update operations can happen at the same time as remove operations, but not before. 

**Lemma 4**:

Assume that all the previous lemmas are correct.

For an "add" operation to be sent to the circuit, there are two possibilities:
1. It's a pure "add" operation
2. It's happening as part of an "update" operation

In case 1, we know from lemma 1 and lemma 2 that the "add" operation must have happened after all "remove" operations.

In case 2, we know from lemma 2 and lemma 3 that "remove" operations happen at the same time or before "update" operations. It follows that "update" operations do not happen _before" remove operations, so "add" operations can't happen before "remove" operations.

Since both cases result in all "remove" operations happening before the first "add" operation, we have shown that the client never adds any root components from a new Blazor endpoint invocation before disposing all components from the previous Blazor endpoint invocation.

QED.
</details>

MackinnonBuck committed:
Enforce component disposal + account for `@key`

MackinnonBuck committed:
Enforce component disposal + account for `@key`

MackinnonBuck commented:
> Otherwise, compare the serialized parameters so even if you pass some arbitrary object, we can still treat it as unchanged if no changed information was going to be passed

I like this a lot! This seems like something we should do. I'm not sure if I'll end up getting time for this, but I'll try if I do.

> [About `@key`]... It seems manageable to me. But what does this do to the implementation cost?

The implementation cost should be fairly small. I've implemented this for Blazor Server - still need to do so for Blazor WebAssembly.

> If that is the case, then using @key will have the same problems, isn't it? Unless we start requiring @key to be unique across the document (essentially making it an id) which is a whole different game.

Since adding a `@key` to a render mode boundary is a new scenario, could we just document this is as the behavior? I'm kind of viewing this as overloading `@key` with two different features rather than `@key` remain a single feature with consistent semantics everywhere.

> Will we avoid the flicker if instead we wait to remove the existing component until we have the HTML for the new component ready and we swap them in one go?

Yep, that should work. It just adds a bit to the implementation, but it might be worth it.

So, following is the remaining work on this PR:
* Implement `@key` support for WebAssembly
* Add unit/E2E tests for everything
* Address feedback
* If time permits, minimize flickering by delaying replacement of interactive component DOM
* If time permits, compare the serialized parameter values instead of their deserialized values

javiercn commented on src/Components/Server/src/Circuits/RootComponentOperationBatch.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Microsoft.AspNetCore.Components.Server;
+
+internal sealed record RootComponentOperationBatch(
```
Use a regular class with a primary constructor.

MackinnonBuck committed:
Improved `@key` handling, shared more code

MackinnonBuck committed:
Improved `@key` handling, shared more code

MackinnonBuck committed:
Max root component count limit

MackinnonBuck committed:
Max root component count limit

MackinnonBuck committed:
No .NET->JS web root component sync anymore

MackinnonBuck committed:
No .NET->JS web root component sync anymore

MackinnonBuck commented:
I've slightly changed how SSR'd components map to interactive components. The previous implementation relied on the client being able to associate SSR'd components with interactive component IDs. This was problematic because:
1. Previously, we needed to be careful not to attempt to send an "update" or "remove" root component operation from JS to .NET until the "add" operation completes and .NET replies with the interactive component ID.
2. Now, with the "update" operation also potentially removing and re-adding the component, we would have needed the client to delay further "update" or "remove" operations until .NET replied with the interactive component ID.

Overall, it's a bit fragile to need to keep the interactive component ID in sync between JS and .NET. Therefore, I've updated the implementation so the JS logic only has to care about its own "SSR component ID" that gets assigned as soon as the SSR'd component gets discovered. It can send root component operations in terms of the SSR component ID instead of the interactive component ID, meaning we don't have to wait for "add" or "update" operations to complete before we do further "update" or "remove" operations.

The fact that "remove" operations could have been delayed actually made invalid my "Lemma 2" in "Proof 2" in my previous comment (which stated that "remove" operations happen as soon as a new endpoint invocation starts). This new change now makes "lemma 2" true, since either:
* We remove a component before we've even sent it to .NET yet, so we can simply stop tracking it without doing further work, or...
* we send a "remove" operation to .NET using the synchronously allocated "SSR component ID".

However, per a discussion offline, rather than going with the approach outlined in my previous comment, we've decided to allow customers to configure a limit on the number of root components originating from a Blazor endpoint. This is a more reliable mechanism whose correctness we can be more confident in.

MackinnonBuck commented:
Remaining work:
* Automated tests
* Further PR feedback
* Eliminate flickering when an "update" causes the disposal/re-instantiation of a root component

MackinnonBuck committed:
Avoid flicker when replacing component instance

MackinnonBuck committed:
Avoid flicker when replacing component instance

MackinnonBuck committed:
Update InteractivityTest.cs

MackinnonBuck committed:
Update InteractivityTest.cs

MackinnonBuck committed:
Fix build issues

MackinnonBuck committed:
Fix build issues

MackinnonBuck committed:
Small improvement

MackinnonBuck committed:
Small improvement

MackinnonBuck committed:
Compare serialized parameter values

MackinnonBuck committed:
Compare serialized parameter values

MackinnonBuck committed:
Cleanup

MackinnonBuck committed:
Cleanup

MackinnonBuck committed:
More cleanup

MackinnonBuck committed:
More cleanup

MackinnonBuck committed:
Create BoundaryMarkerKeyTest.cs

MackinnonBuck committed:
Create BoundaryMarkerKeyTest.cs

MackinnonBuck committed:
WebRootComponentParameters tests

MackinnonBuck committed:
WebRootComponentParameters tests

MackinnonBuck committed:
More interactivity tests

MackinnonBuck committed:
More interactivity tests

MackinnonBuck commented on src/Components/Shared/src/WebRootComponentManager.cs:
```diff
@@ -0,0 +1,175 @@
+using System.Globalization;
+using static Microsoft.AspNetCore.Internal.LinkerFlags;
+
+#if COMPONENTS_SERVER
+namespace Microsoft.AspNetCore.Components.Server.Circuits;
+
+using Renderer = RemoteRenderer;
+
+internal partial class RemoteRenderer
+#elif COMPONENTS_WEBASSEMBLY
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+using Renderer = WebAssemblyRenderer;
+
+internal partial class WebAssemblyRenderer
+#else
+#error WebRootComponentManager cannot be defined in this assembly.
+#endif
+{
+    private WebRootComponentManager? _webRootComponentManager;
+
+    public WebRootComponentManager GetOrCreateWebRootComponentManager()
+        => _webRootComponentManager ??= new(this);
+
+    // Manages components that get added, updated, or removed in Blazor Web scenarios
+    // via Blazor endpoint invocations.
+    public sealed class WebRootComponentManager(Renderer renderer)
```
I know this is a bit unusual. Normally, we would likely put this type of functionality in a base class, but we're trying to avoid changes to public API as much as possible. This class adds a `WebRootComponentManager` class to `RemoteRenderer` and `WebAssemblyRenderer` to manage components rendered by Blazor endpoints. It's a nested class so it can access protected members from each renderer. We also lazily instantiate the `WebRootComponentManager` instance for each renderer so that pure Server or WebAssembly scenarios don't have to pay the cost of the extra bookkeeping.

In the future, we may want to consider pulling this into public API so we don't have to share code in this manner.

javiercn replied:
It's fine to share code in this manner if it's an implementation detail.

MackinnonBuck commented on src/Components/Web.JS/src/Services/WebRootComponentManager.ts:
```diff
@@ -354,67 +358,52 @@ export class WebRootComponentManager implements DescriptorHandler, RootComponent
+      throw new Error(`Could not resolve a root component with SSR component ID '${ssrComponentId}'.`);
     }
 
-    this._pendingComponentsToResolve.delete(selectorId);
-
-    if (component.interactiveComponentId !== undefined) {
-      throw new Error('Cannot resolve a root component for the same descriptor multiple times.');
-    }
-
-    component.interactiveComponentId = componentId;
-
-    // The descriptor may have changed since the last call to handleUpdatedRootComponentsCore().
-    // We'll update this single descriptor so that the component receives the most up-to-date parameters
-    // or gets removed if it no longer exists on the page.
-    this.refreshRootComponents([component]);
-
```
These are the types of things we no longer need to worry about by allowing .NET to control the mapping between SSR'd components and interactive components.

SteveSandersonMS replied:
Impressive

MackinnonBuck commented on src/Components/Shared/src/RootComponentOperation.cs:
```diff
@@ -8,14 +8,8 @@ internal sealed class RootComponentOperation
-    public int? ComponentId { get; set; }
+    // The client side ID of the component to perform the operation on.
+    public int SsrComponentId { get; set; }
 
```
I should clarify that there is a concrete motivation for doing this that relates specifically to this PR.

Since components can be removed/reinitialized upon parameter updates, that means the interactive component ID can change after the browser supplies new component parameters. This would require our JS logic to avoid supplying further parameter updates until it heard back from .NET about whether the component ID changed. This is somewhat fragile, so we would want to avoid doing this as possible.

With this improvement, the browser can supply parameter updates or initiate disposal of an interactive component even before .NET sends any data back to JS. This simplified our existing logic quite a bit, since the JS code has to worry less about asynchrony, and it saved us from complicating the logic further by having to account for the interactive component ID changing dynamically after initialization.

MackinnonBuck committed:
WebRootComponentManager tests

MackinnonBuck committed:
WebRootComponentManager tests

javiercn commented on src/Components/Server/src/CircuitRootComponentOptions.cs:
```diff
@@ -18,4 +18,11 @@ public class CircuitRootComponentOptions : IJSComponentConfiguration
     /// <value>Defaults to <c>100</c>.</value>
     public int MaxJSRootComponents { get; set; } = 100;
+
+    /// <summary>
+    /// Gets or sets the maximum number of root components with an interactive server
+    /// render mode that may exist on a circuit at any given time.
+    /// </summary>
+    /// <value>Defaults to <c>1000</c>.</value>
+    public int MaxInteractiveServerRootComponentCount { get; set; } = 1000;
```
@MackinnonBuck we discussed using `MaxJSRootComponents` directly instead. If in the future, we want to remove this limitation then we don't have to "obsolete" anything.

javiercn commented on src/Components/Shared/src/BoundaryMarkerKey.cs:
```diff
@@ -0,0 +1,53 @@
+
+    public override string ToString()
+    {
+        var separatorCount = 2;
+        var length = componentTypeNameHash.Length + sequenceString.Length + formattedComponentKey.Length + separatorCount;
+        return string.Create(length, (componentTypeNameHash, sequenceString, formattedComponentKey), static (destination, state) =>
+        {
+            WritePart(ref destination, state.componentTypeNameHash);
+            WriteSeparator(ref destination);
+            WritePart(ref destination, state.sequenceString);
+            WriteSeparator(ref destination);
+            WritePart(ref destination, state.formattedComponentKey);
+
+            static void WritePart(ref Span<char> destination, ReadOnlyMemory<char> part)
+            {
+                part.Span.CopyTo(destination);
+                destination = destination[part.Length..];
+            }
+
+            static void WriteSeparator(ref Span<char> destination)
+            {
+                destination[0] = Separator;
+                destination = destination[1..];
+            }
+        });
```
Avoid the custom serialization, just JSON serialize it like we do with the rest of the data.

javiercn commented on src/Components/Shared/src/WebRootComponentParameters.cs:
```diff
@@ -0,0 +1,58 @@
+internal readonly struct WebRootComponentParameters(
+    ParameterView parameterView,
+    IReadOnlyList<ComponentParameter> parameterDefinitions,
+    IReadOnlyList<object> serializedParameterValues)
```
Should this be string?

javiercn replied:
` IReadOnlyList<JsonElement>`

MackinnonBuck replied:
This matches the existing pattern we follow here:
https://github.com/dotnet/aspnetcore/blob/86770ce5945ae3b5651030d97af40d7c701c50b9/src/Shared/Components/ServerComponent.cs#L36
But we could probably convert it to an `IReadOnlyList<JsonElement>` for this case

MackinnonBuck replied:
Hm, we actually probably can't do `IReadOnlyList<JsonElement>` without constructing an entirely new list. Do you think it's fine if we leave this as-is?

wtgodbe commented:
@MackinnonBuck if this is still needed it should be retargeted to `release/8.0`

MackinnonBuck committed:
PR feedback

MackinnonBuck committed:
PR feedback

MackinnonBuck committed:
Some cleanup

MackinnonBuck committed:
Some cleanup

MackinnonBuck committed:
Update blazor.web.js

MackinnonBuck committed:
More tests

MackinnonBuck committed:
More tests

MackinnonBuck committed:
Update WebAssemblyRenderer.cs

MackinnonBuck committed:
Update WebAssemblyRenderer.cs

MackinnonBuck committed:
Update blazor.*.js

SteveSandersonMS commented on src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:
```diff
@@ -190,10 +190,19 @@ public ComponentMarker ToMarker(HttpContext httpContext, int sequence, object? k
+        var sequenceString = sequence.ToString(CultureInfo.InvariantCulture);
+
+        var locationHash = $"{componentTypeNameHash}:{sequenceString}";
+        var formattedComponentKey = (componentKey as IFormattable)?.ToString(null, CultureInfo.InvariantCulture) ?? string.Empty;
```
In your comment at https://github.com/dotnet/aspnetcore/pull/50848#issuecomment-1736054282, you mention that:

> An SSR update can only supply parameter updates to an interactive component if a non-null @key was specified

But the code here always coalesces null into `string.Empty`, so it will never be null, will it? Or do nulls some from some other code path?

Or is the rule actually based on the non-emptiness of the key as opposed to its nullness?

MackinnonBuck replied:
Yes, it uses `string.IsNullOrEmpty` [here](https://github.com/dotnet/aspnetcore/blob/ca1388f73fda974e73bf3469f9e9423cf1fe7cce/src/Components/Shared/src/WebRootComponentManager.cs#L152) rather than just a `null` check.

SteveSandersonMS commented on src/Components/Endpoints/test/WebRootComponentParametersTest.cs:
```diff
@@ -0,0 +1,114 @@
+        {
+            // WebRootComponentParameters expects parameter values to be JsonElements.
+            var jsonElement = JsonSerializer.SerializeToElement(parameterValues[i]);
+            parameterValues[i] = jsonElement;
```
This is a bit sketchy as it's duplicating some implementation inside the unit test, meaning that the unit test is not really testing what its name seems to imply.

However looking at the implementation of `WebRootComponentParameters`, it looks like it will actually throw if you ever pass it any value that isn't a `JsonElement`. If that's the case, I think the test pattern here is fine, since there isn't any other possible scenario than `JsonElement`.

But also if that is true, could we have a unit test asserting that all the parameter values *must* be `JsonElement` and it will throw if not? Then we know how it behaves in all cases.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:
```diff
@@ -9,11 +9,12 @@ import { attachToEventDelegator as attachNavigationManagerToEventDelegator } fro
 const sharedTemplateElemForParsing = document.createElement('template');
 const sharedSvgElemForParsing = document.createElementNS('http://www.w3.org/2000/svg', 'g');
-const elementsToClearOnRootComponentRender: { [componentId: number]: LogicalElement } = {};
+const elementsToClearOnRootComponentRender = new Set<LogicalElement>();
```
Is the reason why you no longer index by `componentId` just that you realised we already have the `LogicalElement` and hence you can look up by that instead? In other words, was this an optimization we could have done all along, or is it only possible now because of something changed in this PR?

MackinnonBuck replied:
It's an optimization we could have done all along. But the reason I'm doing it as part of this PR is that we now reuse the "clear element on first render" logic to allow parameter updates via SSR dispose and re-initialize components without creating a flicker effect between the two operations. However, the `WebRootComponentManager` no longer tracks interactive component IDs, so it's just easier to work with elements instead.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:
```diff
@@ -377,6 +381,14 @@ export function isInteractiveRootComponentElement(element: LogicalElement): bool
+function shouldPreserveContentOnInteractiveComponentDisposal(element: LogicalElement): boolean {
+  return element[preserveContentOnDisposalPropname] === true;
+}
+
```
If this mechanism is as simple as it looks from my naive reading of it, this is a pretty neat technique.

SteveSandersonMS commented on src/Components/test/testassets/Components.TestServer/RazorComponents/Components/InteractiveStreamingRenderingComponent.razor:
```diff
@@ -166,10 +181,13 @@ else
+        return "streaming-interactivity" +
+            $"?{nameof(InitialState)}={stateJson}" +
+            $"&{nameof(ShouldStream)}={shouldStream}" +
+            $"&{nameof(DisableKeys)}={disableKeys}";
```
I'm sure it works fine in these particular tests (since you'll have run them), but it's unusual to see this kind of URL construction without URL-encoding the values.

MackinnonBuck replied:
`stateJson` comes from `GetEscapedStateJson()` which calls `Uri.EscapeDataString()`. The other parameters are just `bool`s. Is there something else I should be doing?

SteveSandersonMS replied:
Ah sorry, I didn't see that. Looks correct then.

SteveSandersonMS commented on src/Shared/Components/ComponentMarker.cs:
```diff
@@ -94,3 +94,22 @@ internal struct ComponentEndMarker
+    // The output should be deterministic between endpoint invocations so that the client
+    // can match up component instances between renders.
+    // The current implementation uses the hashed component type name and its render tree
+    // sequence number.
```
It would be slightly more intuitive to me if it used the *parent component's* type name and sequence number, since that uniquely identifies a point in the app's source code. (By parent, I mean the component that rendered the Component rendertree frame).

If this is using the child component's type name and its sequence number in the parent, that could theoretically clash as two different parent types could render the same child type at the same sequence number.

MackinnonBuck replied:
That's an interesting idea! For some reason, I had originally thought that there might be cases where the sequence number resets to `0` within a single component, which would also create ambiguities. I'll look into changing this.

MackinnonBuck replied:
Another approach would be to use SHA256, which is what @javiercn suggested at one point.

javiercn replied:
@MackinnonBuck sequence numbers are reused within "for" loops I believe, also when using "Region"

MackinnonBuck replied:
@javiercn I think that case is okay, because the same component type will be rendered each iteration. The case I'm more worried about is if you had a matching key but mismatching component type. The browser may not know the type of the component it's working with (in Blazor Server at least), so it might accidentally update a component with a mismatching type if the hash is the same.

javiercn replied:
This is a challenge if we use `@key`, because it's easy for devs to use an object id as the key (for example, customer.Id), so we need to make sure that we create something unique enough to make sure that if I render two sets of separate components using `@key` they don't get "merged".

MackinnonBuck replied:
Is the concern that using `@key` could result in two components getting merged, where they wouldn't have if a `@key` was not specified? Even if customers use `@key`, we still use the component type/sequence number when verifying that components match up.

SteveSandersonMS replied:
I'm just trying to explore where it might theoretically go wrong. If you have this component hierarchy:

 * A
   * MyChild (seq=0, key=123, param=val1)

... and then swap to:

 * B
   * MyChild (seq=0, key=123, param=val2)

... then the location hashes for `MyChild` in both cases may be `MyChild:0:123`, so it may decide to retain `MyChild` and update its param from val1 to val2, even though in the developer's mind, these are unrelated instances since they come from completely different parts of their source code.

If the location hashes were `A:0:123` and `B:0:123` then it would not retain.

MackinnonBuck replied:
Another solution could be to somehow hash multiple levels of the component heirarchy (up to some maximum depth probably).

javiercn replied:
Can we "lazily" walk up the ancestor tree and combine the info "incrementally". Meaning:
* The first time we encounter a server root component, we calculate a hash up to the parent at each level (that we store)
* Then if a second component gets added, we calculate the hash from the common ancestor to the unique branch of the new component.

What this means is that if we have a hierarchy
```
A -> B -> C -> D
A -> B -> C -> E
```

Where D and E are SSRed components, the first time we find D, we Compute the hash up to C (for that, we compute the hash up to B and combine it with the info from C, storing all intermediate results (In component state))

When we are about to Render E and walk up the hierarchy, we see that we already have a hash for C, and just combine it with E.

javiercn replied:
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.incrementalhash?view=net-7.0 Might be of use here

MackinnonBuck replied:
@javiercn I like that a lot!

For the sake of minimizing further changes in this PR, do you think we can consider that as a follow-up item for 8.0 GA?

MackinnonBuck replied:
Opened https://github.com/dotnet/aspnetcore/issues/51073 to track this improvement

SteveSandersonMS reviewed (approved):
While there's definitely too much going on here for me to feel confident I could anticipate any strange edge cases, it looks very believable and is certainly clear you've thought this through well. The tests look good too!

javiercn reviewed (approved):
Looks great!

MackinnonBuck committed:
PR feedback
